### PR TITLE
`Communication`: Fix Tagging other Users does not work 

### PR DESCRIPTION
**Pull request by FelberMartin**
This resolves issue #65 

Now using a different API endpoint for searching tag candidates (matching iOS and WebApp behaviour).

**Out of Scope** 
While testing the implementation I found minor bugs / chances to improve the UI/UX, but I will create a separate issue #72  for this. This PR only focusses on the functionality.

### CODE REVIEW

**Review by julian-wls — APPROVED** [review]
Just tested it and it works as expected. Good work!

The only thing I noticed is, that the preview doesn't get rendered after clicking on the preview button. However, this seems like a topic for another PR in my opinion as it is not really related to the topic of this PR and more related to markdown rendering.

Good to be merged from my side!

**Review by TimOrtel — CHANGES_REQUESTED** [review]
Please add a UI test for this tagging behaviour. Also add an integration test that checks that the application can talk to the server properly.

**Comment by FelberMartin** [comment]
> Also add an integration test that checks that the application can talk to the server properly.

@TimOrtel Could you elaborate? I am not sure what you mean by "the application can talk to the server properly". Should this just be a generic test whether the ConversationViewModel can communicate with a MockServer via `requestReload()` or tailored to the tagging behaviour?

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ReplyTextFieldUiTest.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+        val autoCompleteHints = listOf(
+            AutoCompleteCategory(R.string.markdown_textfield_autocomplete_category_users, listOf(
+                AutoCompleteHint("User1", "<User1>", "1"),
+                AutoCompleteHint("User2", "<User2>", "2"),
+                AutoCompleteHint("User3", "<User3>", "3")
+            ))
+        )
+
+        val mockHintProvider = object : ReplyAutoCompleteHintProvider {
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ReplyTextFieldUiTest.kt:48]: I think calling this a stub would be more appropriate.

**Review by TimOrtel — CHANGES_REQUESTED** [review]
Please also add tests that verify:
- Initially the auto complete hints are not displayed 
- when entering text that is not @ then the auto complete hints are not displayed
- when entering an @ the hints are displayed but when removing the @ again the hints are no longer displayed 
- auto completing can be used twice in a row

**Comment by TimOrtel** [comment]
> > Also add an integration test that checks that the application can talk to the server properly.
> 
> @TimOrtel Could you elaborate, please? I am not sure what you mean by "the application can talk to the server properly". Should this just be a generic test whether the ConversationViewModel can communicate with a MockServer via `requestReload()` or should the test somehow be tailored to the tagging behaviour? If so, how?

You can see exemplary integration tests in feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAnswerMessagesE2eTest.kt

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    private val registerService: RegisterService get() = get()
+
+    @Test(timeout = DefaultTestTimeoutMillis)
+    fun `auto-completion suggests users`() {
+        val usernames = (0 until 3).map {
+            "user$it"
+        }
+
+        runTest(timeout = DefaultTimeoutMillis.milliseconds * 4) {
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:35]: Why do you set the timeout to 4 times the default timeout?

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            usernames.map { username ->
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:36]: Why do you need to register users first? When creating courses there should already be some default users in the course. You can checkout the other end to end tests.

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/reply/ReplyTextFieldUiTest.kt [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    private val autoCompleteHints = listOf(
+        AutoCompleteCategory(
+            R.string.markdown_textfield_autocomplete_category_users, listOf(
+            AutoCompleteHint("User1", "<User1>", "1"),
+            AutoCompleteHint("User2", "<User2>", "2"),
+            AutoCompleteHint("User3", "<User3>", "3"),
+        ))
+    )
+
+    private val mockHintProvider = object : ReplyAutoCompleteHintProvider {
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/reply/ReplyTextFieldUiTest.kt:43]: I think stub is more fitting than fake.

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+
+@Category(EndToEndTest::class)
+@RunWith(RobolectricTestRunner::class)
+class ConversationAutoCompletionE2eTest : ConversationBaseTest() {
+
+    private val registerService: RegisterService get() = get()
+
+    @Test(timeout = DefaultTestTimeoutMillis)
+    fun `auto-completion suggests users`() {
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:30]: Please give your test a name like the other names. For example "test GIVEN users are registered in a course WHEN requesting auto complete users THEN the registered users are returned"

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/reply/ReplyTextFieldUiTest.kt [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        composeTestRule.onNodeWithTag(TEST_TAG_MARKDOWN_TEXTFIELD).assertTextEquals("<User1>")
+        composeTestRule.assertAllAutoCompletionHintsHidden()
+
+        composeTestRule.onNodeWithTag(TEST_TAG_MARKDOWN_TEXTFIELD).performTextInput("@")
+        composeTestRule.assertAllAutoCompletionHintsShown()
+    }
+
+    @Test
+    fun `test GIVEN the textField WHEN entering a first and surname separated by a single whitespace THEN the dialog shows`() {
+        composeTestRule.onNodeWithTag(TEST_TAG_MARKDOWN_TEXTFIELD).performTextInput("@Olaf Scholz")
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ui/reply/ReplyTextFieldUiTest.kt:123]: Please use names of people that do not actually exist.

**Review by TimOrtel — CHANGES_REQUESTED** [review]

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            User(username = user2Username),
+            User(username = user3Username)
+        )
+
+        runTest(timeout = DefaultTimeoutMillis.milliseconds) {
+
+            val typedText = "@user"
+            val autoCompleteSuggestions = conversationService.searchForCourseMembers(
+                courseId = course.id!!,
+                query = typedText,
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:40]: Does this query really work when prepending with the @?

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import kotlin.test.assertEquals
+import kotlin.time.Duration.Companion.milliseconds
+
+
+@Category(EndToEndTest::class)
+@RunWith(RobolectricTestRunner::class)
+class ConversationAutoCompletionE2eTest : ConversationBaseTest() {
+
+    @Test(timeout = DefaultTestTimeoutMillis)
+    fun `test GIVEN users are registered in a course WHEN requesting auto complete users THEN the registered users are returned`() {
```
Comment (TimOrtel) [inline feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:28]: I know the end to end tests right now do not run in the CI, but were you able to verify that this test runs locally?

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt (hunk 7 above)
Comment (FelberMartin) [reply feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:40]: You are write, this probably does not work, I'll adjust it

File: feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt (hunk 8 above)
Comment (FelberMartin) [reply feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:28]: That's actually the problem, the error in the CI and locally were the same, so the tests also do not run locally.  #66 should take care of this and I'll take a deeper look into it as soon as possible next week.

Comment (TimOrtel) [reply feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:28]: The tests should run locally as long as you have an Artemis instance running listening on port 80. You do not need to execute them inside docker.

Comment (FelberMartin) [reply feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:28]: Good to know, I'll try that 👍

Comment (FelberMartin) [reply feature/metis/conversation/src/test/kotlin/de/tum/informatics/www1/artemis/native_app/feature/metis/conversation/ConversationAutoCompletionE2eTest.kt:28]: Sorry for the delay. I tested it with the new docker containers and the test passed

**Review by TimOrtel — APPROVED** [review]
The code looks good to me.

**Pull request merged**